src/compiler/checker.ts

```diff
@@ -18714,6 +18714,18 @@ namespace ts {
             }
         }
 
+        function checkGeneratorWithReturnStatementMustHaveYield (func: FunctionLikeDeclaration) {
+            let hasReturnStmt = false;
+            forEachReturnStatement(<Block>func.body, _ => hasReturnStmt = true);

```

Comment:
seems weird that we have to do this walk again. we have already walked the yield statements to get the types a bit earlier. we can just report the error there instead.